File: clojure/strain/src/strain.clj
(ns strain)

(defn retain [fn coll]
  ;; Note: can be implemented using for with :when modifier,
  ;; but I feel like this is cheating because it's still using
  ;; something that's provided by the standard library.
  (loop [[f & r :as all] coll
         result []]
    (if (empty? all)
      result
      (recur r (if (fn f) (conj result f) result)))))

(defn discard [fn coll]
  (retain (complement fn) coll))

File: javascript/strain/strain.js
export const keep = (collection, predicate) => {
  const result = [];
  for (const element of collection) {
    if (predicate(element)) {
      result.push(element);
    }
  }
  return result;
};

export const discard = (collection, predicate) => {
  return keep(collection, (element) => !predicate(element));
};
